arbnode/espresso_transaction_queue.go

```diff
@@ -0,0 +1,117 @@
+	"slices"
+	"sync"
+	"time"
+
+	espressoClient "github.com/EspressoSystems/espresso-sequencer-go/client"
+	espressoTypes "github.com/EspressoSystems/espresso-sequencer-go/types"
+	"github.com/ethereum/go-ethereum/log"
+	"github.com/offchainlabs/nitro/util/stopwaiter"
+	flag "github.com/spf13/pflag"
+)
+
+type EspressoTransactionQueue struct {
+	stopwaiter.StopWaiter
+
+	config EspressoTransactionQueueConfigFetcher
+
+	espressoClient *espressoClient.Client
+	newTxnNotifier chan struct{}
+
+	pendingTxnsQueueMutex sync.Mutex // cannot be acquired while reorgMutex is held
+	pendingTxnsQueue      []*espressoTypes.Transaction
+
+	submittedTxn *espressoTypes.Transaction
+}
+
+type EspressoTransactionQueueConfig struct {
+	HotShotUrl                  string        `koanf:"hotshot-url"`
+	EspressoNamespace           uint64        `koanf:"espresso-namespace"`
+	EspressoTxnsPollingInterval time.Duration `koanf:"espresso-polling-interval"`
+}
+
+type EspressoTransactionQueueConfigFetcher func() *EspressoTransactionQueueConfig
+
+var DefaultEspressoTransactionQueueConfig = EspressoTransactionQueueConfig{
+	HotShotUrl:                  "",
+	EspressoTxnsPollingInterval: time.Millisecond * 100,
+}
+
+var TestEspressoTransactionQueueConfig = EspressoTransactionQueueConfig{
+	HotShotUrl:                  "",
+	EspressoTxnsPollingInterval: time.Millisecond,
+}
+
+func EspressoTransactionQueueConfigAddOptions(prefix string, f *flag.FlagSet) {
+	f.String(prefix+".hotshot-url", DefaultEspressoTransactionQueueConfig.HotShotUrl, "url of the hotshot sequencer")
+	f.Uint64(prefix+".espresso-namespace", DefaultEspressoTransactionQueueConfig.EspressoNamespace, "espresso namespace that corresponds the L2 chain")
+	f.Duration(prefix+".espresso-polling-interval", DefaultEspressoTransactionQueueConfig.EspressoTxnsPollingInterval, "interval between polling for transactions to be included in the block")
+}
+
+func NewEspressoTransactionQueue(config EspressoTransactionQueueConfigFetcher) *EspressoTransactionQueue {
+
+	espressoClient := espressoClient.NewClient(config().HotShotUrl)
+
+	return &EspressoTransactionQueue{
+		espressoClient:        espressoClient,
+		pendingTxnsQueue:      make([]*espressoTypes.Transaction, 0),
+		pendingTxnsQueueMutex: sync.Mutex{},
+		config:                config,
+	}
+}
+
+func (q *EspressoTransactionQueue) processEspressoTransactions(ctx context.Context, ignored struct{}) time.Duration {
+
+	if q.submittedTxn != nil {
+		// TODO: not sure if this is the correct block height or not arbitrum vs espresso
+		latestBlock, err := q.espressoClient.FetchLatestBlockHeight(ctx)
+		if err != nil {
+			log.Error("failed to fetch latest espresso block height", "err", err)
+			return q.config().EspressoTxnsPollingInterval
+		}
+		// check if each transaction is included in the block and then remove it from the queue
+		fetchedTransactions, err := q.espressoClient.FetchTransactionsInBlock(context.Background(), latestBlock, q.config().EspressoNamespace)
```

Comment:
@ImJeremyHe this currently doesn't verify any proof right? If yes, suggest to add a comment about this so we don't forget.